ttnn/ttnn/operations/binary.py

```diff
@@ -164,8 +164,19 @@ def add(
 
     """
     input_tensor_a = input_tensor_a.value
+    run_with_dtype = False
+    if dtype is not None:
+        if isinstance(input_tensor_b, ttnn.Tensor):
+            run_with_dtype = True
+        else:
+            raise TypeError("ttnn.add: cannot change dtype when broadcasting")
     input_tensor_b = input_tensor_b.value if isinstance(input_tensor_b, ttnn.Tensor) else input_tensor_b
-    output = ttnn._ttnn.operations.binary.add(input_tensor_a, input_tensor_b, memory_config=memory_config, dtype=dtype)
+    if run_with_dtype:
+        output = ttnn._ttnn.operations.binary.add(
+            input_tensor_a, input_tensor_b, memory_config=memory_config, dtype=dtype
+        )
+    else:
+        output = ttnn._ttnn.operations.binary.add(input_tensor_a, input_tensor_b, memory_config=memory_config)
```

Comment:
Update the C++ function that is overloaded for scalar to take in `dtype` and add the assert there